### ISSUE TITLE
Invalid access of discarded (invalid, deleted) uWS.HttpResponse/SSLHttpResponse

### ISSUE THREAD

**Issue by nikosk93** [dalisoft/nanoexpress#28]
Here's a demo file for this error. It's just lists the files of a directory

```javascript
const path = require('path')
const fs = require('fs')
const nanoexpress = require('nanoexpress')
const util = require('util')
const fsReadDir = util.promisify(fs.readdir)

const app = nanoexpress()
const port = 4000

// list the files of `storage/media` directory
app.get('/', async (req, res) => {
  let files = []

  try {
    files = await fsReadDir(path.resolve(__dirname, '../../storage/media'))
    return res.send({error: false, files})
  } catch (err) {
    return res.send({ error: true })
  }
})

// start server
;(async () => {
  try {
    await app.listen(port, '0.0.0.0')
  } catch (err) {
    console.error(err)
  }
})()

```

The problem is that every time that i run and the server and send a `GET /` request, the server throws these warnings

```
(node:5728) UnhandledPromiseRejectionWarning: Invalid access of discarded (invalid, deleted) uWS.HttpResponse/SSLHttpResponse.
(node:5728) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:5728) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

The problem happens when i call the `fsReadDir`. If i move the `res.send` before it, there wont be any warning.

**Comment by dalisoft** [comment]
Hi @nikosk93 

What result do you get if change some lines?

```diff
const path = require('path')
const fs = require('fs')
const nanoexpress = require('nanoexpress')
const util = require('util')
const fsReadDir = util.promisify(fs.readdir)

const app = nanoexpress()
const port = 4000

// list the files of `storage/media` directory
- app.get('/', async (req, res) => {
+ app.get('/', async () => {
   let files = []

  try {
  files = await fsReadDir(path.resolve(__dirname, '../../storage/media'))
-   return res.send({error: false, files})
+   return {error: false, files}
  } catch (err) {
-   return res.send({ error: true })
+   return {error: true}
  }
})

// start server
;(async () => {
  try {
    await app.listen(port, '0.0.0.0')
  } catch (err) {
    console.error(err)
  }
})()
```

Also, i am added new [example](https://github.com/dalisoft/nanoexpress/blob/master/examples/async-folder.js) of your case.

For me, these changes fixed issue and works good.

Thanks for reporting.
Waiting feedback from you!

Have a nice day!

**Comment by nikosk93** [comment]
It works without throwing any warning 👍

Thanks for your quick response and this great project!

**Comment by dalisoft** [comment]
Thank you to reporting.
These issues appears lack of documentation and i will try to fix these issues and documentation when can find time.

**Comment by aprilcoskun** [comment]
@dalisoft I got same issue with following code. I was trying middlewares.
```javascript
const nanoexpress = require('nanoexpress');

function one(req, res, next) {
  req.one = true;
  next();
}

function two(req, res, next) {
  req.two = true;
  next();
}

nanoexpress()
  .use(one, two)
  .get('/favicon.ico', _ => {})
  .get('/', (req, res) => res.send('Hello'))
  .get('/user/:id', (req, res) => {
    res.end(`User: ${req.params.id}`);
  })
  .listen(3000);
```

**Comment by dalisoft** [comment]
Hi @aprilcoskun and thanks for interesting and using this library. 

This may be due of my cache mechanism or maybe issue relates to uWS and I don’t know exactly what happened currently, but i am will try to debug tomorrow, today cannot because of i am out of computer. 
I tomorrow test, inspect and if possible i will fix this, if cannot, will add to docs about this bug. 

Thanks for reporting, this helps this library improve quality and reduce bugs

**Comment by dalisoft** [comment]
@aprilcoskun 
You can try now current changes as workaround

```diff
const nanoexpress = require('nanoexpress');

function one(req, res, next) {
  req.one = true;
  next();
}

function two(req, res, next) {
  req.two = true;
  next();
}

nanoexpress()
  .use(one, two)
- .get('/favicon.ico', _ => {})
+ .get('/favicon.ico', async () => {})
- .get('/', (req, res) => res.send('Hello'))
+ .get('/', (req, res) => res.end('Hello'))
  .get('/user/:id', (req, res) => {
    res.end(`User: ${req.params.id}`);
  })
  .listen(3000);
```

**Comment by dalisoft** [comment]
There even says error like this
`UnhandledPromiseRejectionWarning: Invalid access of discarded (invalid, deleted) uWS.HttpResponse/SSLHttpResponse`, response is responding/returning, but error does not disappearing, but i tried and seems cannot fix this issue, but i will add this to documentation as Known Issues

**Comment by dalisoft** [comment]
Update [docs](https://github.com/dalisoft/nanoexpress/blob/b8203c66021bc7e57b06bb1f2d16489414afdee0/docs/routes.md#known-issues) and i keep this project opened

**Comment by dalisoft** [comment]
Related issue [here](https://github.com/uNetworking/uWebSockets.js/issues/123)

**Comment by aprilcoskun** [comment]
@dalisoft I realized this caused when I try to reach `req.params`. Works fine when I don't call request object, It's looks like request object got deleted somewhere.

**Comment by dalisoft** [comment]
@aprilcoskun Thanks for info, i will research why it happens and if can, i will fix it

**Comment by Ganuzaz** [comment]
I added res.onAborted before res.send/res.end and the warnings gone

the problem is because the uWebSocket
"Every HttpResponse MUST have an attached abort handler IF you do not respond to it immediately inside of the callback. Returning from an Http request handler without attaching (by calling onAborted) an abort handler is ill-use and will termiante. When this event emits, the response has been aborted and may not be used."
https://unetworking.github.io/uWebSockets.js/generated/interfaces/httpresponse.html#onaborted

the same problem from uWebSockets github https://github.com/uNetworking/uWebSockets.js/issues/123

**Comment by dalisoft** [comment]
@Ganuzaz Thanks, it is already here https://github.com/dalisoft/nanoexpress/blob/master/src/handler/http.js#L91

**Comment by Ganuzaz** [comment]
@dalisoft warning still pops out when using sequelize to query.

**Comment by dalisoft** [comment]
@Ganuzaz, sorry, I don’t why it is happens, but PR is welcome

**Comment by dalisoft** [comment]
This was fixed in [**PRO**](https://github.com/dalisoft/nanoexpress/tree/pro) version. In Simple/Free version PR is welcome.

**Comment by dalisoft** [comment]
@nikosk93 This issue was fixed in `v1.1.1` for Simple version.
Sorry for late delay and thanks for your patience!
